python/lsst/ts/mtaos/model.py

```diff
@@ -1059,46 +1063,43 @@ async def _poll_butler_outputs(
         """
         self.log.debug("Polling butler for WEP outputs.")
 
-        butler = dafButler.Butler(self.data_path)
+        butler = dafButler.Butler(self.data_path, collections=[run_name])
         start_time = time.time()
+        elapsed_time = 0.0
 
-        while True:
-            elapsed_time = time.time() - start_time
-
+        while elapsed_time < timeout:
             try:
-                datasetRefs = list(
-                    butler.registry.queryDatasets(
-                        datasetType="postISRCCD", collections=[run_name]
-                    )
+                self.log.info(
+                    f"Querying datasets: zernike_table_name={self.zernike_table_name}, "
+                    f"{run_name=} {extra_id=}."
                 )
-
                 data_ids = butler.registry.queryDatasets(
                     self.zernike_table_name,
                     collections=[run_name],
+                    where=f"visit in ({extra_id})",
                 )
+                if data_ids.any():
+                    self.log.debug(f"Query returned {data_ids.count()} results.")
+                    break
+                else:
+                    self.log.debug("Query returned empty. Continuing.")
             except Exception:
-                self.log.debug(f"Collection '{run_name}' not found")
-                continue
-
-            if data_ids:
-                self.log.debug(f"Found dataset for zernike estimates: {data_ids}")
-                for ref in datasetRefs:
-                    self.log.debug(ref.dataId)
-                break
-
-            if elapsed_time > timeout:
-                raise TimeoutError(
-                    f"Timeout: Could not find outputs for run '{run_name}'"
-                    f" and visit id {visit_id} within {timeout} seconds."
+                self.log.exception(
+                    f"Collection '{run_name}' not found. Waiting {poll_interval}s."
                 )
-
-            self.log.debug(
-                f"Dataset not available yet. Waiting {poll_interval} seconds before retrying..."
+                continue
+            finally:
+                await asyncio.sleep(poll_interval)
+                elapsed_time = time.time() - start_time
+        else:
+            self.log.error(f"Polling loop timed out {timeout=}s, {elapsed_time=}s.")
+            raise TimeoutError(
+                f"Timeout: Could not find outputs for run '{run_name}' "
+                f"and visit id {extra_id} within {timeout} seconds."
             )
-            await asyncio.sleep(poll_interval)
 
         self.log.debug(
-            f"run_name: {run_name}, visit_id: {visit_id} yielded: {data_ids}"
+            f"run_name: {run_name}, visit_id: {extra_id} yielded: {data_ids}"
         )
 
         return [
```

Comment:
A few comments: first and foremost, because these are actually dataset references, you can just `butler.get()` them, and therefore this whole thing can be reduced to simply:
```
return [dRef.dataId['detector'], butler.get(dRef), for dRef in dRefs]
```

Also though, I'd personally replace it with something like this
```
detectorZernikeTuples = [(dRef.dataId["detector"], butler.get(dRef)) for dRef in dRefs]
self.log.info(
    f"Found {len(detectorZernikeTuples)} zernikes for detectors {sorted([d[0] for d in detectorZernikeTuples])}"
)
return detectorZernikeTuples
```
Because I find that having large/complicated listcomps in return statements makes for harder debugging, personally.